src/groovy/org/pih/warehouse/api/StockMovement.groovy

```diff
@@ -225,18 +225,18 @@ class StockMovement {
 
     Boolean isEditAuthorized(Location currentLocation) {
         boolean isSameOrigin = origin?.id == currentLocation?.id
+        boolean isSameDestination = destination?.id == currentLocation?.id
         boolean isDepot = origin?.isDepot()
         boolean isCentralPurchasingEnabled = currentLocation?.supports(ActivityCode.ENABLE_CENTRAL_PURCHASING)
 
-        return !hasBeenReceived() && !hasBeenPartiallyReceived() && (isSameOrigin || !isDepot || !isPending() || isElectronicType() || isCentralPurchasingEnabled)
+        return !hasBeenReceived() && !hasBeenPartiallyReceived() && (isSameOrigin || (!isDepot && isSameDestination) || !isPending() || isElectronicType() || (isCentralPurchasingEnabled && isFromOrder))
```

Comment:
I don't like this line either. We are getting too big and unreadable lines with these checks. I'll merge it for now, but this should be refactored.